src/plugins/intellisense/index.tsx

```diff
@@ -0,0 +1,512 @@
+
+export interface BoundIdentifierFunction {
+  exprId: string;
+  variableName: string;
+  type: "function";
+  id: number;
+  params: string[];
+}
+
+export function getMQCursorPosition(focusedMQ: MathQuillField) {
+  return getController(
+    focusedMQ
+  ).cursor?.cursorElement?.getBoundingClientRect();
+}
+
+const intellisenseMountPoint = document.createElement("div");
+document.body.appendChild(intellisenseMountPoint);
+
+export function getSelectedExpressionID(): string | undefined {
+  return Object.keys(Calc.controller.listModel.selectedItemMap)[0];
+}
+
+export default class Intellisense extends PluginController {
+  static id = "intellisense" as const;
+  static enabledByDefault = true;
+
+  view: MountedComponent | undefined;
+
+  x: number = 0;
+  y: number = 0;
+
+  intellisenseOpts: BoundIdentifier[] = [];
+  intellisenseIndex: number = -1;
+
+  latestIdent: TryFindMQIdentResult | undefined;
+  latestMQ: MathQuillField | undefined;
+
+  intellisenseReturnMQ: MathQuillField | undefined;
+  prevCursorPos: { x: number; y: number } | undefined;
+  goRightBeforeReturningToMQ: boolean = false;
+
+  idcounter = 0;
+
+  partialFunctionCall: PartialFunctionCall | undefined;
+  partialFunctionCallIdent: BoundIdentifier | undefined;
+  partialFunctionCallDoc: string | undefined;
+
+  intellisenseState = new IntellisenseState(getMetadata());
+
+  canHaveIntellisense = false;
+
+  // recalculate the intellisense
+  updateIntellisense() {
+    const focusedMQ = MathQuillView.getFocusedMathquill();
+    this.saveCursorState();
+    this.intellisenseOpts = [];
+
+    // is there actually a focused mathquill window?
+    if (focusedMQ) {
+      // find the identifier the cursor is at
+      this.latestIdent = getMathquillIdentifierAtCursorPosition(focusedMQ);
+      if (this.latestIdent)
+        this.latestIdent.ident = this.latestIdent.ident.replace(/ /g, "");
+
+      this.latestMQ = focusedMQ;
+
+      // determine if the user is in a partial function call
+      this.partialFunctionCall = getPartialFunctionCall(focusedMQ);
+      this.partialFunctionCallIdent = Array.from(
+        this.intellisenseState.boundIdentifiers()
+      ).find(
+        (i) =>
+          addBracketsToIdent(i.variableName) ===
+            this.partialFunctionCall?.ident && i.type === "function"
+      );
+
+      // if the user is in a partial function call,
+      // find its documentation if it exists
+      const models = Calc.controller.getAllItemModels();
+      let found = false;
+      for (let i = 0; i < models.length; i++) {
+        const current = models[i];
+        if (
+          this.partialFunctionCallIdent &&
+          current.type === "text" &&
+          models[i + 1]?.type === "expression" &&
+          this.partialFunctionCallIdent.exprId === models[i + 1]?.id
+        ) {
+          this.partialFunctionCallDoc = current.text;
+          found = true;
+        }
+      }
+      if (!found) this.partialFunctionCallDoc = undefined;
```

Comment:
Could this be `this.partialFunctionCallDoc = models.find((current, i) => conditionHere)`? It's automatically undefined if not found.

Reply:
Fixed!